Mathlib/Analysis/NormedSpace/Star/ContinuousFunctionalCalculus/Instances.lean

```diff
@@ -0,0 +1,269 @@
+Authors: Jireh Loreaux
+-/
+import Mathlib.Analysis.NormedSpace.Star.ContinuousFunctionalCalculus.Restrict
+import Mathlib.Analysis.NormedSpace.Star.Spectrum
+import Mathlib.Analysis.NormedSpace.Star.ContinuousFunctionalCalculus
+
+/-! # Instances of the continuous functional calculus
+
+## Main definitions
+
+* `IsStarNormal.instContinuousFunctionalCalculus`: the continuous functional calculus for normal
+  elements in a unital C⋆-algebra over `ℂ`.
+* `IsSelfAdjoint.instContinuousFunctionalCalculus`: the continuous functional calculus for
+  selfadjoint elements in a unital C⋆-algebra over `ℂ`.
+* `Nonneg.instContinuousFunctionalCalculus`: the continuous functional calculus for nonnegative
+  elements in a unital C⋆-algebra over `ℂ`, which is also a `StarOrderedRing`.
+
+## Tags
+
+continuous functional calculus, normal, selfadjoint
+-/
+
+noncomputable section
+
+section Normal
+
+instance IsStarNormal.cfc_map {R A : Type*} {p : A → Prop} [CommSemiring R] [StarRing R]
+    [MetricSpace R] [TopologicalSemiring R] [ContinuousStar R] [TopologicalSpace A] [Ring A]
+    [StarRing A] [Algebra R A] [ContinuousFunctionalCalculus R p] (a : A) (f : R → R) :
+    IsStarNormal (cfc a f) where
+  star_comm_self := by
+    rw [Commute, SemiconjBy]
+    by_cases h : ContinuousOn f (spectrum R a)
+    · rw [← cfc_star, ← cfc_mul .., ← cfc_mul ..]
+      congr! 2
+      exact mul_comm _ _
+    · simp [cfc_apply_of_not_continuousOn a h]
+
+variable {A : Type*} [NormedRing A] [StarRing A] [CstarRing A] [CompleteSpace A]
+variable [NormedAlgebra ℂ A] [StarModule ℂ A]
+
+instance IsStarNormal.instContinuousFunctionalCalculus :
+    ContinuousFunctionalCalculus ℂ (IsStarNormal : A → Prop) where
+  exists_cfc_of_predicate a ha := by
+    refine ⟨(elementalStarAlgebra ℂ a).subtype.comp <| continuousFunctionalCalculus a,
+      ?hom_closedEmbedding, ?hom_id, ?hom_map_spectrum, ?predicate_hom⟩
+    case hom_closedEmbedding => exact Isometry.closedEmbedding <|
+      isometry_subtype_coe.comp <| StarAlgEquiv.isometry (continuousFunctionalCalculus a)
+    case hom_id => exact congr_arg Subtype.val <| continuousFunctionalCalculus_map_id a
+    case hom_map_spectrum =>
+      intro f
+      simp only [StarAlgHom.comp_apply, StarAlgHom.coe_coe, StarSubalgebra.coe_subtype]
+      rw [← StarSubalgebra.spectrum_eq (elementalStarAlgebra.isClosed ℂ a),
+        AlgEquiv.spectrum_eq (continuousFunctionalCalculus a), ContinuousMap.spectrum_eq_range]
+    case predicate_hom => exact fun f ↦ ⟨by rw [← map_star]; exact Commute.all (star f) f |>.map _⟩
+
+lemma IsSelfAdjoint.spectrumRestricts {a : A} (ha : IsSelfAdjoint a) :
+    SpectrumRestricts a Complex.reCLM where
+  rightInvOn _x hx := ha.mem_spectrum_eq_re hx |>.symm
+  left_inv := Complex.ofReal_re
+
+/-- An element in a C⋆-algebra is selfadjoint if and only if it is normal and its spectrum is
+contained in `ℝ`. -/
+lemma isSelfAdjoint_iff_isStarNormal_and_spectrumRestricts {a : A} :
```

Comment:
Is it possible to relax the `CstarRing A` assumption for this and some of the other results below (i.e. `mem_unitary_of_spectrum_subset_circle`? It would be nice to be able to use those results for e.g. matrices.

Reply:
Yes, I can see how that would be convenient. However, I'm not really sure how to do it without moving certain facts used in the proof out as hypotheses (although I can do that if you want).

For instance, `isSelfAdjoint_iff_isStarNormal_and_spectrumRestricts` uses, in the forward direction, the fact that selfadjoint elements have real spectrum, which depends in turn on `spectrum.subset_circle_of_unitary` as well as `NormedSpace.exp`. I think if we had a class that states `‖u‖ = 1` for `u` unitary, then we might be able to reduce the requirements to basically `NormedSpace.exp`, but barring that I don't see a way to incorporate both the abstract situation and matrices in a single theorem.

Reply:
ah yes, indeed it is possible. When I wrote this originally I had considered it, but then thought I couldn't get around using some `CstarRing` properties somewhere. However, you can, the key is `eqOn_of_cfc_eq_cfc`. Note that this is one of reasons we need the full power of the functional calculus (here we're using injectivity).

I have implemented these changes so that it will apply to matrices.